subsys/bluetooth/controller/hci/hci.c

```diff
@@ -4798,6 +4800,151 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+
+static void vs_err_fatal_cpu_data_fill(bt_hci_vs_fatal_error_cpu_data *cpu_data,
+				       const z_arch_esf_t *esf)
+{
+	cpu_data->a1 = sys_cpu_to_le32(esf->basic.a1);
+	cpu_data->a2 = sys_cpu_to_le32(esf->basic.a2);
+	cpu_data->a3 = sys_cpu_to_le32(esf->basic.a3);
+	cpu_data->a4 = sys_cpu_to_le32(esf->basic.a4);
+	cpu_data->ip = sys_cpu_to_le32(esf->basic.ip);
+	cpu_data->lr = sys_cpu_to_le32(esf->basic.lr);
+	cpu_data->xpsr = sys_cpu_to_le32(esf->basic.xpsr);
+}
+#endif /* CONFIG_CPU_CORTEX_M */
+
+static struct net_buf *vs_err_evt_create(uint8_t subevt, uint8_t len)
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&vs_err_tx_pool, K_FOREVER);
+	if (buf) {
+		struct bt_hci_evt_le_meta_event *me;
+		struct bt_hci_evt_hdr *hdr;
+
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_EVT);
+
+		hdr = net_buf_add(buf, sizeof(*hdr));
+		hdr->evt = BT_HCI_EVT_VENDOR;
+		hdr->len = len + sizeof(*me);
+
+		me = net_buf_add(buf, sizeof(*me));
+		me->subevent = subevt;
+	}
+
+	return buf;
+}
+
+struct net_buf *hci_vs_err_stack_frame(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* Prepare vendor specific HCI Fatal Error event */
+	struct bt_hci_vs_fatal_error_stack_frame *sf;
+	bt_hci_vs_fatal_error_cpu_data *cpu_data;
+	struct net_buf *buf;
+
+	buf = vs_err_evt_create(BT_HCI_EVT_VS_ERROR_DATA_TYPE_STACK_FRAME,
+				sizeof(*sf) + sizeof(*cpu_data));
+	if(buf != NULL) {
+		sf = net_buf_add(buf, (sizeof(*sf) + sizeof(*cpu_data)));
+		sf->reason = sys_cpu_to_le32(reason);
+		sf->cpu_type = BT_HCI_EVT_VS_ERROR_CPU_TYPE_CORTEX_M;
+
+		vs_err_fatal_cpu_data_fill(
+			(bt_hci_vs_fatal_error_cpu_data *)sf->cpu_data, esf);
+	} else {
+		BT_ERR("Can't create HCI Fatal Error event");
+	}
+
+	return buf;
+}
+
+static struct net_buf *hci_vs_err_trace_create(uint8_t data_type,
+					       const char *file_path,
+					       uint32_t line, uint64_t pc)
+{
+	uint32_t len = 0U, pos = 0U;
+	struct net_buf *buf = NULL;
+
+	if (file_path) {
+		/* Extract file name from a path */
+		while (file_path[len] != '\0') {
+			if (file_path[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
+		file_path += pos;
+		len -= pos;
+
+		/* If file name was found in file_path, in other words: file_path is not empty
+		 * string and is not `foo/bar/`.
+		 */
+		if (len) {
+			/* Total data length: len = file name strlen + \0 + sizeof(line number)
+			 * Maximum length of an HCI event data is BT_BUF_RX_SIZE. If total data
+			 * length exceeds this maximum, truncate file name.
+			 */
+			uint32_t data_len = 1 + sizeof(line);
+
+			/* If a buffer is created for a TRACE data, include sizeof(pc) in total
+			 * length.
+			 */
+			if (data_type == BT_HCI_EVT_VS_ERROR_DATA_TYPE_TRACE) {
+				data_len += sizeof(pc);
+			}
+
+			if (data_len + len > BT_BUF_RX_SIZE) {
+				uint32_t overflow_len = len + data_len - BT_BUF_RX_SIZE;
+
+				/* Truncate the file name length by number of overflow bytes */
+				data_len = data_len + (len - overflow_len);
```

Comment:
Isn't `len` the length of the file name (perhaps we need to rename it to be more explicit?)? 

You are still doing `net_buf_add_mem(buf, file_path, len);`, and since we are not reducing `len` here, it seems that we are still adding the entire file path (do we need the file path, and not just the file name?).

Reply:
I thought about rename it, but decided to leave it as is... but it would be better to have an explicit name like `file_name_len`. I'll update it.

`net_buf_add_mem(buf, file_path, file_name_len);` is correct, because the `file_name_len` if updated to store correct file name length after end of `while` loop that searches for the file name:
```
/* Extract file name from a path */
while (file_path[file_name_len] != '\0') {
	if (file_path[file_name_len] == '/') {
		pos = file_name_len + 1;
	}
	file_name_len++;
}
file_path += pos;
file_name_len -= pos; // <--------- here full path length is updated to store file name length. pos stores position of last '/' + 1
```

Reply:
You were right, there were an issue with `file_name_len` it was not subtracted with `overflow_len` before `net_buf_add_mem(buf, file_path, file_name_len);` 
What more, the `data_len` was updated with `file_name_len` only in case of overflow. The `data_len += file_name_len;` should be done after `if` statement that checks the overflow.

Both issues are fixed now.